aiohttp/connector.py

```diff
@@ -477,35 +490,37 @@ def _get(self, key):
         del self._conns[key]
         return None
 
-    def _release_key_waiter(self, key, waiters):
-        if not waiters:
-            return False
+    def _release_waiter(self):
+        """
+        Iterates over all waiters till found one that is not finsihed and
+        belongs to a host that has available connections.
+        """
+        if not self._waiters:
+            return
 
-        waiter = waiters.popleft()
-        if not waiter.done():
-            waiter.set_result(None)
+        # Having the dict keys ordered this avoids to iterate
+        # at the same order at each call.
+        queues = list(self._waiters.keys())
+        random.shuffle(queues)
```

Comment:
Why do we need shuffling?
The operation is not free.

Reply:
Yes, that hurts me. But without the shuffle the iteration of the queues will be done always in the same order,  having chances of starvation for certain queues.

We could try to get rid of that operation, but it will imply some important changes.

Reply:
The idea of @thehesiod of having a "true" FIFO would get rid of this operation, but my gut feeling says that this FIFO queue will have enough footprint - has to take care of the limit per host - to make it slower than just a call `shuffle`.

PD: a shuffle operation takes around 1 microsecond.

Reply:
`waiters` is a `deque`.
`deque.rotate()` is cheap. Can we use it?